### PR TITLE
Add dedicated region service

### PR DESCRIPTION
**Pull request by FingerLeader**
### Check List 

- Possible to add a test?
- Don't forget to run `go generate` if you modify `examples`.

### CODE REVIEW

File: tidbcloud/dedicated_api_client.go [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+		return nil, err
+	}
+	return &DedicatedClientDelegate{
+		dc: dc,
+		ic: ic,
+	}, nil
+}
+
+func (d *DedicatedClientDelegate) ListRegions(ctx context.Context) (*dedicated.TidbCloudOpenApidedicatedv1beta1ListRegionsResponse, error) {
+	tflog.Debug(ctx, fmt.Sprintf("dc.cfg: %v", *d.dc.GetConfig()))
```
Comment (zhangyangyu) [inline tidbcloud/dedicated_api_client.go:77]: why this is Debug level and others are Trace level?

File: internal/provider/provider.go [hunk 2]
```diff
@@ -108,9 +113,27 @@ func (p *tidbcloudProvider) Configure(ctx context.Context, req provider.Configur
...
+		iamEndpoint = os.Getenv(TiDBCloudIAMEndpoint)
+	}
+	dc, err := NewDedicatedClient(publicKey, privateKey, dedicatedEndpoint, iamEndpoint, fmt.Sprintf("%s/%s", UserAgent, p.version))
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Unable to create client",
+			"Unable to create tidb dedicated client:\n\n"+err.Error(),
+		)
+		return
+	}
```
Comment (zhangyangyu) [inline internal/provider/provider.go:132]: 1. I think we don't need to use the default endpoint here, just throw env value in and hide the logic inside the client. 
2. Why dedicated client need IAM endpoint? I think there should be a IAM client.
3. tidb dedicated -> TiDB Cloud Dedicated; plz make sure messages user can see is in formal format.

File: tidbcloud/dedicated_api_client.go [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+	resp, h, err := d.dc.RegionServiceAPI.RegionServiceListRegions(ctx).Execute()
+	tflog.Trace(ctx, fmt.Sprintf("ListRegions: %v, h: %v, err: %v", resp, h, err))
+	return resp, parseError(err, h)
+}
+
+func (d *DedicatedClientDelegate) GetRegion(ctx context.Context, regionId string) (*dedicated.Commonv1beta1Region, error) {
+	resp, h, err := d.dc.RegionServiceAPI.RegionServiceGetRegion(ctx, regionId).Execute()
```
Comment (zhangyangyu) [inline tidbcloud/dedicated_api_client.go:84]: and here no log?

File: examples/data-sources/tidbcloud_dedicated_region/data-source.tf [hunk 4]
```diff
@@ -0,0 +1,11 @@
+variable "region_id" {
+  type = string
```
Comment (zhangyangyu) [inline examples/data-sources/tidbcloud_dedicated_region/data-source.tf:2]: The default value for nullable is true.  Why here omit nullable and other two data source have.

File: internal/provider/dedicated_cloud_providers_data_source.go [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+			fmt.Sprintf("Error in Configure: expected %T but got %T", tidbcloudProvider{}, req.ProviderData))
+	}
+}
+
+func (d *dedicatedCloudProvidersDataSource) Schema(_ context.Context, _ datasource.SchemaRequest, resp *datasource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		MarkdownDescription: "dedicated cloud providers data source",
+		Attributes: map[string]schema.Attribute{
+			"project_id": schema.StringAttribute{
+				MarkdownDescription: "The ID of the project.",
```
[the remark this answers is not in the document]
Comment (zhangyangyu) [reply internal/provider/dedicated_cloud_providers_data_source.go:48]: +1

File: tidbcloud/dedicated_api_client.go [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+func parseError(err error, resp *http.Response) error {
+	if resp != nil {
+		defer resp.Body.Close()
+	}
+	if err == nil {
+		return nil
+	}
+	if resp == nil {
+		return err
+	}
```
[the remark this answers is not in the document]
Comment (FingerLeader) [reply tidbcloud/dedicated_api_client.go:103]: Updated this function as same as `parseError()` in tidbcloud-cli

**Pull request merged**